File: kypa_calc/household.py
"""Build PolicyEngine household situations for KYPA calculations."""

_GROUP_UNITS = ["families", "spm_units", "tax_units", "households"]


def _add_member_to_units(situation, member_id):
    """Append a member to all group units (families, spm, tax, households)."""
    for unit in _GROUP_UNITS:
        key = next(iter(situation[unit]))
        situation[unit][key]["members"].append(member_id)


def build_household_situation(
    age_head: int,
    age_spouse: int | None,
    dependent_ages: list[int],
    year: int = 2026,
    with_axes: bool = False,
    max_earnings: int = 2_000_000,
    state_code: str = "CA",
) -> dict:
    """
    Build a PolicyEngine household situation dict.

    Args:
        age_head: Age of household head.
        age_spouse: Age of spouse (None if single).
        dependent_ages: List of dependent ages.
        year: Tax year.
        with_axes: If True, adds AGI sweep axis.
        max_earnings: Maximum AGI for the sweep axis.
        state_code: State code.

    Returns:
        PolicyEngine situation dict.
    """
    situation = {
        "people": {"you": {"age": {year: age_head}}},
        "families": {"your family": {"members": ["you"]}},
        "marital_units": {"your marital unit": {"members": ["you"]}},
        "spm_units": {"your household": {"members": ["you"]}},
        "tax_units": {"your tax unit": {"members": ["you"]}},
        "households": {
            "your household": {
                "members": ["you"],
                "state_code": {year: state_code},
            }
        },
    }

    if with_axes:
        situation["axes"] = [
            [
                {
                    "name": "adjusted_gross_income",
                    "min": 0,
                    "max": max_earnings,
                    "count": min(4_001, max(501, max_earnings // 500)),
                    "period": year,
                    "target": "tax_unit",
                }
            ]
        ]

    if age_spouse is not None:
        situation["people"]["your partner"] = {
            "age": {year: age_spouse}
        }
        _add_member_to_units(situation, "your partner")
        situation["marital_units"]["your marital unit"][
            "members"
        ].append("your partner")

    for i, dep_age in enumerate(dependent_ages):
        if i == 0:
            child_id = "your first dependent"
        elif i == 1:
            child_id = "your second dependent"
        else:
            child_id = f"dependent_{i + 1}"

        situation["people"][child_id] = {"age": {year: dep_age}}
        _add_member_to_units(situation, child_id)
        situation["marital_units"][f"{child_id}'s marital unit"] = {
            "members": [child_id]
        }

    return situation


def calculate_household_impact(params: dict) -> dict:
    """Run baseline and reform simulations and return impact data."""
    from policyengine_us import Simulation
    from .reforms import create_reform

    year = params.get("year", 2026)
    reform = create_reform(year=year)

    situation = build_household_situation(
        age_head=params["age_head"],
        age_spouse=params.get("age_spouse"),
        dependent_ages=params.get("dependent_ages", []),
        year=year,
        with_axes=True,
        max_earnings=int(params.get("max_earnings", 500_000)),
        state_code=params.get("state_code", "CA"),
    )

    sim_baseline = Simulation(situation=situation)
    sim_reform = Simulation(situation=situation, reform=reform)

    net_bl = sim_baseline.calculate("household_net_income", period=year)
    net_rf = sim_reform.calculate("household_net_income", period=year)
    diff = net_rf - net_bl

    agi_range = sim_baseline.calculate(
        "adjusted_gross_income", period=year, map_to="household"
    )

    income = params.get("income", 75000)
    idx = int((abs(agi_range - income)).argmin())

    return {
        "income_range": agi_range.tolist(),
        "net_income_change": diff.tolist(),
        "benefit_at_income": {
            "baseline": float(net_bl[idx]),
            "reform": float(net_rf[idx]),
            "difference": float(diff[idx]),
        },
        "x_axis_max": int(params.get("max_earnings", 500_000)),
    }

File: frontend/lib/household.ts
/**
 * Build a PolicyEngine household situation for the PE API.
 *
 * Mirrors kypa_calc/household.py but in TypeScript so we can call
 * the PE API directly from the frontend (no backend needed).
 */

import type { HouseholdRequest } from "./types";

const GROUP_UNITS = ["families", "spm_units", "tax_units", "households"] as const;

function addMemberToUnits(
  situation: Record<string, any>,
  memberId: string
): void {
  for (const unit of GROUP_UNITS) {
    const key = Object.keys(situation[unit])[0];
    situation[unit][key].members.push(memberId);
  }
}

export function buildHouseholdSituation(
  params: HouseholdRequest
): Record<string, any> {
  const {
    age_head,
    age_spouse,
    dependent_ages,
    income,
    year,
    max_earnings,
    state_code,
  } = params;
  const yearStr = String(year);
  const axisMax = Math.max(max_earnings, income);

  const situation: Record<string, any> = {
    people: { you: { age: { [yearStr]: age_head }, employment_income: { [yearStr]: null } } },
    families: { "your family": { members: ["you"] } },
    marital_units: { "your marital unit": { members: ["you"] } },
    spm_units: { "your household": { members: ["you"] } },
    tax_units: {
      "your tax unit": {
        members: ["you"],
        adjusted_gross_income: { [yearStr]: null },
      },
    },
    households: {
      "your household": {
        members: ["you"],
        state_code: { [yearStr]: state_code },
        household_net_income: { [yearStr]: null },
      },
    },
    axes: [
      [
        {
          name: "employment_income",
          min: 0,
          max: axisMax,
          count: Math.min(4001, Math.max(501, Math.floor(axisMax / 500))),
          period: yearStr,
          target: "person",
        },
      ],
    ],
  };

  if (age_spouse != null) {
    situation.people["your partner"] = { age: { [yearStr]: age_spouse } };
    addMemberToUnits(situation, "your partner");
    situation.marital_units["your marital unit"].members.push("your partner");
  }

  for (let i = 0; i < dependent_ages.length; i++) {
    const childId =
      i === 0
        ? "your first dependent"
        : i === 1
          ? "your second dependent"
          : `dependent_${i + 1}`;

    situation.people[childId] = { age: { [yearStr]: dependent_ages[i] } };
    addMemberToUnits(situation, childId);
    situation.marital_units[`${childId}'s marital unit`] = {
      members: [childId],
    };
  }

  return situation;
}

/**
 * Build the KYPA reform policy dict for the PE API.
 *
 * Uses the full POLICY_JSON from kypa_calc/reforms.py, translated to
 * the PE API format: { parameter_name: { "period.period": value } }.
 */
export function buildReformPolicy(): Record<string, Record<string, any>> {
  return {
    "gov.contrib.congress.afa.in_effect": {
      "2026-01-01.2100-12-31": true,
    },
    "gov.contrib.congress.afa.ctc.amount.base": {
      "2026-01-01.2026-12-31": 3600,
      "2027-01-01.2027-12-31": 3720,
      "2028-01-01.2028-12-31": 3840,
      "2029-01-01.2029-12-31": 3960,
      "2030-01-01.2031-12-31": 4080,
      "2032-01-01.2032-12-31": 4200,
      "2033-01-01.2034-12-31": 4320,
      "2035-01-01.2035-12-31": 4440,
    },
    "gov.irs.credits.eitc.eligibility.age.max": {
      "2026-01-01.2100-12-31": 200,
    },
    "gov.irs.credits.eitc.eligibility.age.min": {
      "2026-01-01.2100-12-31": 19,
    },
    "gov.irs.credits.eitc.eligibility.age.min_student": {
      "2026-01-01.2100-12-31": 24,
    },
    "gov.irs.credits.eitc.max[0].amount": {
      "2026-01-01.2026-12-31": 1502,
      "2027-01-01.2027-12-31": 1577,
      "2028-01-01.2028-12-31": 1610,
      "2029-01-01.2029-12-31": 1645,
      "2030-01-01.2030-12-31": 1678,
      "2031-01-01.2031-12-31": 1712,
      "2032-01-01.2032-12-31": 1746,
      "2033-01-01.2033-12-31": 1779,
      "2034-01-01.2034-12-31": 1815,
      "2035-01-01.2035-12-31": 1851,
    },
    "gov.irs.credits.eitc.phase_in_rate[0].amount": {
      "2026-01-01.2100-12-31": 0.153,
    },
    "gov.irs.credits.eitc.phase_out.rate[0].amount": {
      "2026-01-01.2100-12-31": 0.153,
    },
    "gov.irs.credits.eitc.phase_out.start[0].amount": {
      "2026-01-01.2026-12-31": 11610,
      "2027-01-01.2027-12-31": 12190,
      "2028-01-01.2028-12-31": 12440,
      "2029-01-01.2029-12-31": 12710,
      "2030-01-01.2030-12-31": 12970,
      "2031-01-01.2031-12-31": 13230,
      "2032-01-01.2032-12-31": 13490,
      "2033-01-01.2033-12-31": 13750,
      "2034-01-01.2034-12-31": 14030,
      "2035-01-01.2035-12-31": 14300,
    },
    "gov.irs.deductions.standard.amount.SINGLE": {
      "2026-01-01.2026-12-31": 37500,
      "2027-01-01.2027-12-31": 38450,
      "2028-01-01.2028-12-31": 39250,
      "2029-01-01.2029-12-31": 40100,
      "2030-01-01.2030-12-31": 40950,
      "2031-01-01.2031-12-31": 41750,
      "2032-01-01.2032-12-31": 42600,
      "2033-01-01.2033-12-31": 43400,
      "2034-01-01.2034-12-31": 44250,
      "2035-01-01.2035-12-31": 45150,
    },
    "gov.irs.deductions.standard.amount.JOINT": {
      "2026-01-01.2026-12-31": 75000,
      "2027-01-01.2027-12-31": 76900,
      "2028-01-01.2028-12-31": 78500,
      "2029-01-01.2029-12-31": 80200,
      "2030-01-01.2030-12-31": 81900,
      "2031-01-01.2031-12-31": 83500,
      "2032-01-01.2032-12-31": 85200,
      "2033-01-01.2033-12-31": 86800,
      "2034-01-01.2034-12-31": 88500,
      "2035-01-01.2035-12-31": 90300,
    },
    "gov.irs.deductions.standard.amount.HEAD_OF_HOUSEHOLD": {
      "2026-01-01.2026-12-31": 56250,
      "2027-01-01.2027-12-31": 57700,
      "2028-01-01.2028-12-31": 58900,
      "2029-01-01.2029-12-31": 60200,
      "2030-01-01.2030-12-31": 61400,
      "2031-01-01.2031-12-31": 62650,
      "2032-01-01.2032-12-31": 63900,
      "2033-01-01.2033-12-31": 65150,
      "2034-01-01.2034-12-31": 66400,
      "2035-01-01.2035-12-31": 67750,
    },
    "gov.irs.deductions.standard.amount.SEPARATE": {
      "2026-01-01.2026-12-31": 37500,
      "2027-01-01.2027-12-31": 38450,
      "2028-01-01.2028-12-31": 39250,
      "2029-01-01.2029-12-31": 40100,
      "2030-01-01.2030-12-31": 40950,
      "2031-01-01.2031-12-31": 41750,
      "2032-01-01.2032-12-31": 42600,
      "2033-01-01.2033-12-31": 43400,
      "2034-01-01.2034-12-31": 44250,
      "2035-01-01.2035-12-31": 45150,
    },
    "gov.irs.deductions.standard.amount.SURVIVING_SPOUSE": {
      "2026-01-01.2026-12-31": 75000,
      "2027-01-01.2027-12-31": 76900,
      "2028-01-01.2028-12-31": 78500,
      "2029-01-01.2029-12-31": 80200,
      "2030-01-01.2030-12-31": 81900,
      "2031-01-01.2031-12-31": 83500,
      "2032-01-01.2032-12-31": 85200,
      "2033-01-01.2033-12-31": 86800,
      "2034-01-01.2034-12-31": 88500,
      "2035-01-01.2035-12-31": 90300,
    },
    "gov.contrib.congress.afa.ctc.phase_out.threshold.lower.JOINT": {
      "2026-01-01.2026-12-31": 150000,
    },
    "gov.contrib.congress.afa.ctc.phase_out.threshold.lower.SINGLE": {
      "2026-01-01.2026-12-31": 112500,
    },
    "gov.contrib.congress.afa.ctc.phase_out.threshold.lower.SURVIVING_SPOUSE": {
      "2026-01-01.2026-12-31": 150000,
    },
    "gov.contrib.congress.afa.ctc.phase_out.threshold.lower.HEAD_OF_HOUSEHOLD": {
      "2026-01-01.2026-12-31": 112500,
    },
    // Top income tax rate increases: 35% → 41%, 37% → 43%
    "gov.irs.income.bracket.rates.6": {
      "2026-01-01.2100-12-31": 0.41,
    },
    "gov.irs.income.bracket.rates.7": {
      "2026-01-01.2100-12-31": 0.43,
    },
  };
}

/**
 * Linear interpolation helper — find the value at `x` in sorted arrays.
 */
export function interpolate(
  xs: number[],
  ys: number[],
  x: number
): number {
  if (x <= xs[0]) return ys[0];
  if (x >= xs[xs.length - 1]) return ys[ys.length - 1];
  for (let i = 1; i < xs.length; i++) {
    if (xs[i] >= x) {
      const t = (x - xs[i - 1]) / (xs[i] - xs[i - 1]);
      return ys[i - 1] + t * (ys[i] - ys[i - 1]);
    }
  }
  return ys[ys.length - 1];
}
